google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java

```diff
@@ -253,14 +253,16 @@ public long executePartitionedUpdate(final Statement stmt, final UpdateOption...
 
   private <T> T runWithSessionRetry(Function<Session, T> callable) {
     PooledSessionFuture session = getSession();
-    while (true) {
-      try {
-        return callable.apply(session);
-      } catch (SessionNotFoundException e) {
-        session =
-            (PooledSessionFuture)
-                pool.getPooledSessionReplacementHandler().replaceSession(e, session);
+    try {
+      while (true) {
+        try {
+          return callable.apply(session);
+        } catch (SessionNotFoundException e) {
+          session = pool.getPooledSessionReplacementHandler().replaceSession(e, session);
+        }
       }
+    } finally {
+      session.close();
```

Comment:
The responsibility to close the session at the end has been moved from the underlying method that is being called to this method. This prevents the `PooledSession` from being closed twice if the session was replaced. That would again cause the `numSessionsInUse` to be decreased by 1 more than it should, which could cause it to become negative.